consensus/consensus.go

```diff
@@ -97,6 +97,10 @@ type Engine interface {
 	FinalizeAndAssemble(chain ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction,
 		uncles []*types.Header, receipts []*types.Receipt) (*types.Block, error)
 
+	// FinalizeAtIndex implements consensus.Engine, accumulating the block and uncle rewards,
+	// setting the final state on the header at an index.
+	FinalizeAtIndex(chain ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction, uncles []*types.Header, index int)
```

Comment:
I'm questioning the need for all of these "AtIndex" functions. They all duplicate a bunch of logic and open up the possibility of unintended differences between implementations. Each should probably just use variatic args to allow optional specifying of a node context